File: app/src/main/java/com/worldventures/dreamtrips/modules/tripsimages/service/command/EditPhotoWithTagsCommand.java
package com.worldventures.dreamtrips.modules.tripsimages.service.command;

import com.innahema.collections.query.queriables.Queryable;
import com.worldventures.dreamtrips.R;
import com.worldventures.dreamtrips.core.api.action.CommandWithError;
import com.worldventures.dreamtrips.core.janet.dagger.InjectableAction;
import com.worldventures.dreamtrips.modules.common.model.UploadTask;
import com.worldventures.dreamtrips.modules.common.view.custom.tagview.viewgroup.newio.model.PhotoTag;
import com.worldventures.dreamtrips.modules.tripsimages.model.Photo;
import com.worldventures.dreamtrips.modules.tripsimages.service.TripImagesInteractor;

import java.util.List;

import javax.inject.Inject;

import io.techery.janet.command.annotations.CommandAction;
import rx.Observable;

@CommandAction
public class EditPhotoWithTagsCommand extends CommandWithError<Photo> implements InjectableAction {

   private String uid;
   private UploadTask task;
   private List<PhotoTag> addedTags;
   private List<PhotoTag> removedTags;

   @Inject TripImagesInteractor tripImagesInteractor;

   public EditPhotoWithTagsCommand(String uid, UploadTask task, List<PhotoTag> addedTags, List<PhotoTag> removedTags) {
      this.uid = uid;
      this.task = task;
      this.addedTags = addedTags;
      this.removedTags = removedTags;
   }

   @Override
   protected void run(CommandCallback callback) throws Throwable {
      tripImagesInteractor.editPhotoActionPipe().createObservableResult(new EditPhotoCommand(uid, task))
            .map(editPhotoCommand -> {
               Photo photo = editPhotoCommand.getResult();
               addedTags.removeAll(photo.getPhotoTags());
               return photo;
            })
            .flatMap(photo -> {
               if (addedTags.isEmpty()) return Observable.just(photo);
               return tripImagesInteractor.addPhotoTagsActionPipe()
                     .createObservableResult(new AddPhotoTagsCommand(uid, addedTags))
                     .map(result -> photo);
            })
            .flatMap(photo -> {
               if (removedTags.isEmpty()) return Observable.just(photo);
               List<Integer> userIds = Queryable.from(removedTags)
                     .concat(photo.getPhotoTags())
                     .map(photoTag -> photoTag
                     .getUser()
                     .getId()).toList();
               return tripImagesInteractor.deletePhotoTagsPipe()
                     .createObservableResult(new DeletePhotoTagsCommand(uid, userIds))
                     .map(result -> photo);
            })
            .doOnNext(photo -> {
               photo.getPhotoTags().addAll(addedTags);
               photo.getPhotoTags().removeAll(removedTags);
               photo.setPhotoTagsCount(photo.getPhotoTags().size());
            })
            .subscribe(callback::onSuccess, callback::onFail);
   }

   @Override
   public int getFallbackErrorMessage() {
      return R.string.photo_edit_error;
   }
}

File: app/src/main/java/com/worldventures/dreamtrips/social/ui/tripsimages/service/command/EditPhotoWithTagsCommand.kt
package com.worldventures.dreamtrips.social.ui.tripsimages.service.command

import com.worldventures.core.janet.CommandWithError
import com.worldventures.dreamtrips.R
import com.worldventures.dreamtrips.modules.common.model.UploadTask
import com.worldventures.dreamtrips.social.ui.feed.view.custom.tagview.viewgroup.newio.model.PhotoTag
import com.worldventures.dreamtrips.social.ui.tripsimages.model.Photo
import com.worldventures.dreamtrips.social.ui.tripsimages.service.TripImagesInteractor
import com.worldventures.janet.injection.InjectableAction
import io.techery.janet.Command
import io.techery.janet.command.annotations.CommandAction
import rx.Observable
import javax.inject.Inject

@CommandAction
class EditPhotoWithTagsCommand(private val uid: String, private val task: UploadTask, private val addedTags: MutableList<PhotoTag>,
                               private val removedTags: List<PhotoTag>) : CommandWithError<Photo>(), InjectableAction {

   @Inject internal lateinit var tripImagesInteractor: TripImagesInteractor

   @Throws(Throwable::class)
   override fun run(callback: Command.CommandCallback<Photo>) {
      tripImagesInteractor.editPhotoActionPipe.createObservableResult(EditPhotoCommand(uid, task))
            .map {
               val photo = it.result
               addedTags.removeAll(photo.photoTags)
               photo
            }
            .flatMap(this::getAddPhotoTagsObservable)
            .flatMap(this::getRemovePhotoTagsObservable)
            .doOnNext { photo ->
               photo.photoTags.addAll(addedTags)
               photo.photoTags.removeAll(removedTags)
               photo.photoTagsCount = photo.photoTags.size
            }
            .subscribe(callback::onSuccess, callback::onFail)
   }

   private fun getAddPhotoTagsObservable(photo: Photo) = if (addedTags.isEmpty()) {
      Observable.just(photo)
   } else tripImagesInteractor.addPhotoTagsActionPipe
         .createObservableResult(AddPhotoTagsCommand(uid, addedTags))
         .map { photo }

   private fun getRemovePhotoTagsObservable(photo: Photo) = if (removedTags.isEmpty()) {
      Observable.just(photo)
   } else tripImagesInteractor.deletePhotoTagsPipe
         .createObservableResult(DeletePhotoTagsCommand(uid, removedTags.plus(photo.photoTags).map { it.user.id }))
         .map { photo }

   override fun getFallbackErrorMessage() = R.string.photo_edit_error
}
